### PR TITLE
Thread2

### PR DESCRIPTION
**Pull request by Lucarda**
Improvements to the threading part

### CODE REVIEW

File: flite.c [hunk 1]
```diff
@@ -148,50 +147,82 @@ static void flite_synth(t_flite *x) {
...
 #endif
   x->x_wave = flite_text_to_wave(x->x_textbuf, x->x_voice);
 
-  if (!x->x_wave) {
-
+  if (!x->x_wave) 
+  {
     pthread_mutex_lock(&x->x_mutex);
-    if (x->x_requestcode != QUIT)
+    if (!x->x_shutdown)
```
Comment (Spacechild1) [inline flite.c:217]: what's the reason for adding `x_shutdown` instead of using the `QUIT` request?

File: flite.c [hunk 2]
```diff
@@ -618,18 +664,21 @@ static void *flite_new(t_symbol *ary)
 }
 
 static void flite_free(t_flite *x) {
-    
+
+  x->x_shutdown = 1;
   while(x->x_inprogress) {
-  sleep(1);
+    sleep(1);
```
Comment (Spacechild1) [inline flite.c:670]: What's the point of the sleep here? Some lines below you're setting the `QUIT` request, that should do it.

If you have a threading problem, *never* work it around with a `sleep` call! Instead, fix the actual problem :-)

File: flite.c (hunk 1 above)
Comment (Lucarda) [reply flite.c:217]: because we never get to `QUIT` if  `x->x_inprogress`

Comment (Spacechild1) [reply flite.c:217]: well, that's a bug then. better fix the bug instead of adding a workaround.

File: flite.c (hunk 2 above)
Comment (Spacechild1) [reply flite.c:670]: Also, what's the point of `while(x->x_inprogress)`?

**Comment by Spacechild1** [comment]
Also, please check all places where you set `x_inprogress`. You should only touch this variable *on the main thread*: set it to 1 in the Pd method and set it back to 0 in the clock tick function.

Finally, you only have to set `x_inprogress=1` in *asynchronous* methods (where it is set back to 0 in the clock tick function).  E.g. there is no point in doing `x_inprogress = 1` and `x_inprogress = 0` in `flite_synth()` because it will execute *synchronously* on Pd's main thread.

**Comment by Lucarda** [comment]
> There's still a problem with flite_do_textfile because it calls flite_synth, even if called from the worker thread.

done

> Finally, please check all places where you set x_inprogress. You should only touch this variable on the main thread: ...

done

> Finally, I would recommend to rearrange the code a bit: flite_synth and flite_thr_synth really belong together, while flite_thread_synth belongs to flite_thread.

I fail to understand what this is about. You mean a better ordering of the functions?

**Comment by Spacechild1** [comment]
> I fail to understand what this is about. You mean a better ordering of the functions?

yes

**Comment by Lucarda** [comment]
@Spacechild1  I think this is roughly ready to merge, bump to 0.3.2 and do a Deken release.

What do you think?

File: flite.c [hunk 3]
```diff
@@ -126,154 +129,147 @@ typedef struct _flite
...
+  x->x_requestcode = VOXFILE;
   pthread_mutex_unlock(&x->x_mutex);
- 
+  pthread_cond_signal(&x->x_requestcondition);
   return;
 }
 
 /*--------------------------------------------------------------------
- * flite_clock_tick : clock
+ * flite_thrd_voice_file : open the voice file .
```
Comment (Spacechild1) [inline flite.c:251]: Now this is confusing. For the other methods, `flite_thrd_*` is used for the Pd method that notifies the worker thread, while `flite_thread_*` is used for corresponding helper function called in the worker thread.

BTW, what's the reason for only having a multi-threaded method for reading a voice file?

File: flite.c [hunk 4]
```diff
@@ -126,154 +129,147 @@ typedef struct _flite
...
-      pd_error(x,"flite: synthesis failed for text '%s'", x->x_textbuf);
-  } else if (x->x_threaderrormsg == INPROGRESS) {      
-      pd_error(x,"%s", thread_waiting);
+  flite_add_lang("eng",usenglish_init,cmulex_init);
+  flite_add_lang("usenglish",usenglish_init,cmulex_init);
+  x->x_voice = flite_voice_load(x->x_reqfile);
+  pthread_mutex_lock(&x->x_mutex);
+  if (x->x_requestcode != QUIT)
+  {
+    if (x->x_requestcode != IDLE)
```
[the remark this answers is not in the document]
Comment (Spacechild1) [reply flite.c:264]: Single-threaded methods can be very useful, e.g. when running Pd in batch mode.

File: flite.c [hunk 5]
```diff
@@ -468,67 +374,220 @@ static void flite_do_textfile(t_flite *x) {
...
+#ifdef FLITE_DEBUG
+  debug("flite: got message 'synth'\n");
+#endif
+
+  if (!x->x_textbuf) 
+  {
+    pthread_mutex_lock(&x->x_mutex);
+    if (x->x_requestcode != QUIT)
+    {
+      if (x->x_requestcode != IDLE)
```
Comment (Spacechild1) [inline flite.c:459]: see above

File: flite.c [hunk 6]
```diff
@@ -468,67 +374,220 @@ static void flite_do_textfile(t_flite *x) {
...
+  debug("flite: flite_text_to_wave()\n");
+#endif
+  x->x_wave = flite_text_to_wave(x->x_textbuf, x->x_voice);
+
+  if (!x->x_wave) 
+  {
+    pthread_mutex_lock(&x->x_mutex);
+    if (x->x_requestcode != QUIT)
+    {
+      if (x->x_requestcode != IDLE)
```
Comment (Spacechild1) [inline flite.c:481]: see above

File: flite.c (hunk 3 above)
Comment (Spacechild1) [reply flite.c:251]: Synchronous methods can be very useful, e.g. when running Pd in batch mode.

File: flite.c [hunk 7]
```diff
@@ -468,67 +374,220 @@ static void flite_do_textfile(t_flite *x) {
...
+#ifdef FLITE_DEBUG
+  debug("flite: cst_wave_resample()\n");
+#endif
+  cst_wave_resample(x->x_wave, sys_getsr());
   
   pthread_mutex_lock(&x->x_mutex);
-  x->x_requestcode = TEXTFILE;
+  if (x->x_requestcode != QUIT)
+  {
+    if (x->x_requestcode != IDLE)
```
Comment (Spacechild1) [inline flite.c:502]: see above

Comment (Lucarda) [reply flite.c]: done

File: flite.c (hunk 6 above)
Comment (Lucarda) [reply flite.c]: done

File: flite.c (hunk 5 above)
Comment (Lucarda) [reply flite.c]: done

File: flite.c [hunk 8]
```diff
@@ -126,155 +129,175 @@ typedef struct _flite
...
 
 #ifdef FLITE_DEBUG
-  debug("flite: flite_text_to_wave()\n");
+  debug("flite_voice: called with arg='%s'\n", vox->s_name);
 #endif
-  x->x_wave = flite_text_to_wave(x->x_textbuf, x->x_voice);
 
-  if (!x->x_wave) {
+  const char *voxstring;
+  voxstring = vox->s_name;
```
Comment (Spacechild1) [inline flite.c:208]: This is really nitpicky, but why do you have the assignment on a seperate line? This should be simply `const char *voxstring = vox->s_name`

(I guess this comes from a misunderstanding about C. With C89 you had to declare any local variables at the top of the block, which often lead to a split between declaration and assignment. With C99 local variables can be declared anywhere.)

Comment (Spacechild1) [reply flite.c:208]: you don't have to change it, it's more meant as a comment

**Comment by Spacechild1** [comment]
> @Spacechild1 I think this is roughly ready to merge, bump to 0.3.2 and do a Deken release.
> 
> What do you think?

Looks good to me! I haven't really tested it in practice, but I guess you did :-)

**Comment by Lucarda** [comment]
> Looks good to me! I haven't really tested it in practice, but I guess you did :-)

Total thanks @Spacechild1 :)

it's up on Deken.

**Pull request merged**